File: hackerrank/balancedBrackets.js
//www.hackerrank.com/challenges/balanced-brackets/problem?h_l=interview&playlist_slugs%5B%5D=interview-preparation-kit&playlist_slugs%5B%5D=stacks-queues

// Complete the function isBalanced in the editor below. It must return a string: YES if the sequence is balanced or NO if it is not.

https: function isBalanced(s) {
  const bracketMap = {
    "(": ")",
    "{": "}",
    "[": "]",
  };
  const stack = [];
  let result = "YES";
  s.split("").forEach((item) => {
    if (bracketMap[item]) {
      stack.push(bracketMap[item]);
    } else if (stack.pop() !== item) {
      result = "NO";
    }
  });
  if (stack.length) result = "NO";
  return result;
}

console.log(isBalanced("{[()]}")); // Yes
console.log(isBalanced("{{[[(())]]}}")); // Yes
console.log(isBalanced("{[(])}")); // No
console.log(isBalanced("{(([])[])[]}")); // Yes
console.log(isBalanced("{(([])[])[]]}")); // No

File: hackerrank/Python/balancedBrackets.py
# https://www.hackerrank.com/challenges/balanced-brackets/problem?h_l=interview&playlist_slugs%5B%5D=interview-preparation-kit&playlist_slugs%5B%5D=stacks-queues


def isBalanced(s):
    # Complete this function
    bracket_map = {'{': '}', '(': ')', '[': ']'}
    stack = []
    for b in s:
        if b in bracket_map.keys():  # openning
            stack.append(bracket_map[b])
        elif b in bracket_map.values():  # closing
            if not stack or stack.pop() != b:
                return "NO"
    if not stack:
        return "YES"
    else:
        return "NO"


print(isBalanced("{[()]}"))  # Yes
print(isBalanced("{{[[(())]]}}"))  # Yes
print(isBalanced("{[(])}"))  # No
print(isBalanced("{(([])[])[]}"))  # Yes
print(isBalanced("{(([])[])[]]}"))  # No
